### ISSUE TITLE
Interface type explosion causes error on query

### ISSUE THREAD

**Issue by martin-svanberg-northvolt** [apollographql/federation#2935]
**Describe the bug**
Attempting to query an interface with many implementations results in an error.

**To Reproduce**

We have an interface (`FactoryUnit`) with more than 70 implementations. It's defined as follows:

```graphql
# Graph A
interface FactoryUnit {
  id: ID!
  # ...
  unitState: FactoryUnitState
}

type FactoryUnitState {
  id: ID!
  # ...
}
```

We try to run the following query, which gives an error:

Query:
```graphql
query {
	identity(id: "...") {
		id
		... on FactoryUnit { unitState { id } }
	}
}
```

**Expected behavior**
It should not return an error when querying.

**Output**
```json
{
	"errors": [
		{
			"message": "value retrieval failed: router bridge error: the deno runtime raised an error: `request: couldn't receive response couldn't deserialize payload `6476623271848066925`: `deno: couldn't deserialize response : `Error(\"invalid neither null nor empty object: found Object {\\\"code\\\": String(\\\"QUERY_PLANNING_FAILED\\\"), \\\"exception\\\": Object {\\\"stacktrace\\\": Array [String(\\\"RangeError: Invalid array length\\\")]}}\", line: 0, column: 0)``.`",
			"extensions": {
				"code": "INTERNAL_SERVER_ERROR"
			}
		}
	]
}
```

**Desktop (please complete the following information):**
macOS Monterey, but issue also visible on the server running Alpine Linux 3.16

**Additional context**
We've traced this issue to the following function in Apollo Federation, which appears to be a function with exponential time and space complexity: https://github.com/apollographql/federation/blob/92a9dffdc50f6cbc73aa6225d0536d448cb7bf9a/query-graphs-js/src/graphPath.ts#L2167. It fails when trying to construct an array of length 2^70. Given the comment above `flatCartesianProduct`, it seems that this is a known issue.

**Comment by abernix** [comment]
Thanks for opening this.  I'm moving this to the `federation` repository since anything that can be fixed will be in the `federation` repository, rather than in the Router itself.

**Comment by martin-svanberg-northvolt** [comment]
1. This is using Federation 2.3
2. I've verified that all of them are on either 2.0 or 2.3
3. Sure! Give me some time and I'll try to set up a minimal reproduction

**Comment by WeaselScience** [comment]
We are facing the same issue in our schema. We're upgrading from Apollo Gateway 0.16.0 to the latest version of Apollo Router. We have a supergraph that triggers this issue, however, we were unable to create a lean reproduction graph.

We would be eager to share our supergraph that triggers the issue, but, as it's not public, could we do so privately? @abernix

**Comment by namirsab** [comment]
@Meschreiber maybe could you tell us how to deal with this? we are blocked now as we can't upgrade to apollo router and this is our only problem. We want to help with a reproduction as @WeaselScience said.

**Comment by TylerBloom** [comment]
> We are facing the same issue in our schema. We're upgrading from Apollo Gateway 0.16.0 to the latest version of Apollo Router. We have a supergraph that triggers this issue, however, we were unable to create a lean reproduction graph.

@WeaselScience and @namirsab, two questions. 1) Have you tested the Rust query planner in the router? This issue seems to be with the query planner, and the new Rust query planner will very soon be the default planner. It would be helpful to know if this problem also exists in the new planner, or, if not, you should be able to use the new query planner already. 2) Do you use studio? If so, I can view your schema and operations from there. You can either link to it or provide the organization and schema variant's name.

**Comment by WeaselScience** [comment]
@TylerBloom Thank you for your response!

Using the native Rust query planner produces an error that is described in the issue:
```
value retrieval failed: Federation error: An internal error has occurred, please report this bug to Apollo.

Details: flat_cartesian_product: excessive number of combinations: 32614907904
```

I've published the affected project onto Studio. The organization name is `amperecloud`, the graph is `reproduction`, variant `current`.

You can trigger the issue by running the following query:

```
query GetDeviceById($deviceId: ID!) {
  getDeviceById(deviceId: $deviceId) {
    id
    company {
      name
    }
  }
}
```
```
{
  "deviceId": "123"
}
```
...which will produce the error above.

**Comment by dariuszkuc** [comment]
Hello 👋 
Thanks @WeaselScience for providing the repro. 

TLDR -> the reason why query blows up is actually provided in the native query planner error message. Your query actually results in **32 BILLION** combinations in how it can be resolved which results in legacy (JS) QP going OOM and the native (RS) QP rejecting it due to the excessive complexity. 

When doing the query planning we first evaluate all available options in how to construct the query. Any option to take different path to resolve a field will grow those options exponentially. Even though there might be really only a single viable plan, we need to evaluate all options*. This ensures that we can actually plan it and provides a guarantee that generated plan is valid. 

Federation v2 made a number of improvements in how we handle things that removes ambiguity and includes better ownership model. While migrating from v1 composition to v2 composition is often a seamless process (in most cases it should just work), it can result in some additional QP complexity until all yours subgraphs are upgraded to fed v2 as well. It looks like none of your subgraphs are upgraded to v2 yet. Once you upgrade your subgraphs, those queries *should* work. See [our migration docs](https://www.apollographql.com/docs/graphos/reference/migration/to-federation-version-2) for additional details.

---
*we might change this algorithm in the future

**Comment by namirsab** [comment]
@dariuszkuc thanks for your help!

@WeaselScience and I are thinking how to tackle this better. The guideline you've sent us says that we can update the subgraphs once at a time. We are wondering if this is the case also for us. We are wondering where to start, because we have many subgraphs and this will be a big effort. Do you have a suggestion based on your explanation? which subgraph would reduce the amount of combinations that the query planner would produce?

And, just a naive question, is there a way to give the query planner a hint to reduce the combinations?

Thanks in advance!
